order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java

```diff
@@ -37,22 +54,116 @@ public void removeUntrackedItem(String itemName) {
+                throw new IncorrectResponseException("Response is not correct!");
+            }
+
+            Item item = new Item();
+            item.setPrice(tickResponse.getPrice());
+            item.setQuantity(tickResponse.getQuantity());
+            item.setOrigin(tickResponse.getOrigin());
+
+            productsQuantity.merge(tickResponse.getGoodName(), tickResponse.getQuantity(), Long::sum);
+
+            Set<Item> itemSet = allItems.get(tickResponse.getGoodName());
+            if (itemSet != null) {
+                if (itemSet.contains(item)) {
+                    Item presentItem = itemSet.stream().filter(item1 -> Double.compare(item1.getPrice(), item.getPrice()) == 0
+                            && item1.getOrigin().equals(item.getOrigin())).findFirst().get();
+                    presentItem.setQuantity(presentItem.getQuantity() + item.getQuantity());
+                    return;
+                }
+
+                itemSet.add(item);
+            } else {
+                LOGGER.error("Item: {} is not being tracked and cannot be added to itemMarket!",
+                        tickResponse.getGoodName());
+            }
+
+            LOGGER.info("Products data updated!");
+        } else {
+            throw new IncorrectResponseException("Response is not correct!");
+        }
+    }
+
+    public List<OrderBookLayer> getRequestedItem(String itemName, long quantity) {
+        LOGGER.info("Getting requested item: {} with quantity: {}", itemName, quantity);
+        TreeSet<Item> items = this.allItems.get(itemName);
+
+        if (items == null || productsQuantity.get(itemName) < quantity) {
+            return Collections.emptyList();
+        }
+
+        List<OrderBookLayer> layers;
+        try {
+            lock.writeLock().lock();
+
+            layers = new ArrayList<>();
+
+            Iterator<Item> iterator = items.iterator();
+            long itemLeft = quantity;
+
+            while (iterator.hasNext() && itemLeft > 0) {
+                Item currentItem = iterator.next();
+
+                OrderBookLayer.Builder currentLayer = OrderBookLayer.newBuilder()
+                        .setPrice(currentItem.getPrice());
+
+                if (currentItem.getQuantity() >= itemLeft) {
+                    currentLayer.setQuantity(itemLeft);
+
+                    if (currentItem.getQuantity() == itemLeft) {
+                        iterator.remove();
+                    }
+
+                    currentItem.setQuantity(currentItem.getQuantity() - itemLeft);
+                } else if (currentItem.getQuantity() < itemLeft) {
+                    currentLayer.setQuantity(currentItem.getQuantity());
+
+                    iterator.remove();
+                }
+                productsQuantity.put(itemName, productsQuantity.get(itemName) - currentLayer.getQuantity());
+                itemLeft -= currentLayer.getQuantity();
+
+                OrderBookLayer build = currentLayer
```

Comment:
orderBookLayer would be a better name for the object

Reply:
Fixed